Mathlib/Tactic/CC/Datatypes.lean

```diff
@@ -0,0 +1,640 @@
+
+namespace CCState
+
+/-- Get the root representative of the given expression. -/
+def root (ccs : CCState) (e : Expr) : Expr :=
+  match ccs.entries.find? e with
+  | some n => n.root
+  | none => e
+#align cc_state.root Mathlib.Tactic.CC.CCState.root
+
+/-- Get the next element in the equivalence class.
+Note that if the given `Expr` `e` is not in the graph then it will just return `e`. -/
+def next (ccs : CCState) (e : Expr) : Expr :=
+  match ccs.entries.find? e with
+  | some n => n.next
+  | none => e
+#align cc_state.next Mathlib.Tactic.CC.CCState.next
+
+/-- Check if `e` is the root of the congruence class. -/
+def isCgRoot (ccs : CCState) (e : Expr) : Bool :=
+  match ccs.entries.find? e with
+  | some n => e == n.cgRoot
+  | none => true
+#align cc_state.is_cg_root Mathlib.Tactic.CC.CCState.isCgRoot
+
+/--
+"Modification Time". The field `mt` is used to implement the mod-time optimization introduce by the
+Simplify theorem prover. The basic idea is to introduce a counter `gmt` that records the number of
+heuristic instantiation that have occurred in the current branch. It is incremented after each round
+of heuristic instantiation. The field `mt` records the last time any proper descendant of of thie
+entry was involved in a merge. -/
+def mt (ccs : CCState) (e : Expr) : Nat :=
+  match ccs.entries.find? e with
+  | some n => n.mt
+  | none => ccs.gmt
+#align cc_state.mt Mathlib.Tactic.CC.CCState.mt
+
+def inSingletonEqc (ccs : CCState) (e : Expr) : Bool :=
+  match ccs.entries.find? e with
+  | some it => it.next == e
+  | none => true
+#align cc_state.in_singlenton_eqc Mathlib.Tactic.CC.CCState.inSingletonEqc
+
+def getRoots (ccs : CCState) (roots : Array Expr) (nonsingletonOnly : Bool) : Array Expr :=
+  Id.run do
+    let mut roots := roots
+    for (k, n) in ccs.entries do
+      if k == n.root && (!nonsingletonOnly || !ccs.inSingletonEqc k) then
+        roots := roots.push k
+    return roots
+
+def checkEqc (ccs : CCState) (e : Expr) : Bool :=
+  toBool <| Id.run <| OptionT.run do
+    let root := ccs.root e
+    let mut size : Nat := 0
+    let mut it := e
+    repeat
+      let some itN := ccs.entries.find? it | failure
+      guard (itN.root == root)
+      let mut it₂ := it
+      -- following `target` fields should lead to root
+      repeat
+        let it₂N := ccs.entries.find? it₂
+        match it₂N.bind Entry.target with
+        | some it₃ => it₂ := it₃
+        | none => break
+      guard (it₂ == root)
+      it := itN.next
+      size := size + 1
+    until it == e
+    guard (ccs.entries.find? root |>.any (·.size == size))
+
+def checkInvariant (ccs : CCState) : Bool :=
+  ccs.entries.all fun k n => k != n.root || checkEqc ccs k
+
+def getNumROccs (ccs : CCState) (e : Expr) (inLHS : Bool) : Nat :=
+  match ccs.acEntries.find? e with
+  | some ent => (ent.ROccs inLHS).size
+  | none => 0
+
+def getVarWithLeastOccs (ccs : CCState) (e : ACApps) (inLHS : Bool) : Option Expr :=
```

Comment:
What does `var` mean in this context? I assume it's not a `Expr.bvar`/`Expr.fvar`?

Reply:
`Entry.acVar`

Reply:
```suggestion
/-- Search for the AC-variable (`Entry.acVar`) with the least occurrences in the state. -/
def getVarWithLeastOccs (ccs : CCState) (e : ACApps) (inLHS : Bool) : Option Expr :=
```